### DISCUSSION TITLE
Questions about data persistence

### DISCUSSION THREAD

**Discussion by ktamaral** [milvus-io/milvus#36866] [Q&A and General discussion] [answered]
Hello I'm looking for more information on how data persistence works in Milvus.

Specifically, I would like to know more details about what happens when the entire k8s server crashes and all nodes are restarted, resulting in the in-memory storage being cleared out.

Here's my current understanding of how it works:

_The system initially loads data into memory for fast access, and then the data is also written to disk periodically. When the server restarts, Milvus automatically reloads the persisted data from disk back into memory. This process is automatic, allowing continued usage of the database without needing to manually re-ingest data. The settings can be configured for how much data is kept in memory._

Please let me know if these points are correct and please provide links to documentation where I can find more information on the details, if available.

Thank you!

**Comment by yhmo** [comment, marked as answer]
After you insert data into a collection, the data is persisted in S3/minio as "sealed-segment", and the path of the data files are stored in etcd.
After you call collection.create_index(), milvus creates index for each sealed-segment. Index files are also persisted in S3/minio, and the path of the index files are aslo stored in etcd.
When you call collection.load(), the query nodes read the index files from S3/mionio, read the data into memory. If index files are not ready for some sealed segments, the query nodes will read their original data files into memory. With in-memory index, we can get best search performance.
If the milvus is crash/restarted, the query nodes automatically load the collection(which is loaded before) into memory. The data files and index files are always there, on the S3/mino, nothing changed. You don't need to re-ingest the data again. The query nodes just load the data/index files into memory then you are able to search again.
